### PR TITLE
Move certain events to the newly introduced `ServerPlayerEvent`

### PR DESCRIPTION
**Pull request by Spinoscythe**
Some modders in the community noticed that some `PlayerEvent` events, like `PlayerEvent.PlayerLoggedInEvent`, only ever pass in a ServerPlayer instance. However, since `PlayerEvent` is a `Player` type, you have to cast the player back to `ServerPlayer`.

It would be worth switching to a `<? extends Player>` generic type for player events, but we don't support generic events anymore. So this seems the most ideal.

Some modders = @LeScooter. ty for the idea :)
Also @Technici4n approves this idea.
image

### CODE REVIEW

File: src/main/java/net/neoforged/neoforge/event/entity/item/ItemTossEvent.java [hunk 1]
```diff
@@ -21,11 +21,11 @@ public class ItemTossEvent extends ItemEvent implements ICancellableEvent {
...
-     * @param entityItem The EntityItem being tossed.
+     * @param itemEntity The EntityItem being tossed.
      * @param player     The player tossing the item.
      */
-    public ItemTossEvent(ItemEntity entityItem, Player player) {
-        super(entityItem);
+    public ItemTossEvent(ItemEntity itemEntity, Player player) {
+        super(itemEntity);
         this.player = player;
     }
```
Comment (marchermans) [inline src/main/java/net/neoforged/neoforge/event/entity/item/ItemTossEvent.java]: Random parameter rename

**Review by marchermans — CHANGES_REQUESTED** [review]

File: src/main/java/net/neoforged/neoforge/event/entity/item/ItemTossEvent.java (hunk 1 above)
Comment (Spinoscythe) [reply src/main/java/net/neoforged/neoforge/event/entity/item/ItemTossEvent.java]: done!

File: src/main/java/net/neoforged/neoforge/event/entity/player/ServerPlayerEvent.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+package net.neoforged.neoforge.event.entity.player;
+
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.level.GameType;
+import net.minecraft.world.level.Level;
+import net.neoforged.bus.api.ICancellableEvent;
+import net.neoforged.neoforge.event.entity.living.LivingEvent;
+
+public abstract class ServerPlayerEvent extends LivingEvent {
```
Comment (marchermans) [inline src/main/java/net/neoforged/neoforge/event/entity/player/ServerPlayerEvent.java]: Would extending `PlayerEvent` not make more sense, giving that `ServerPlayer` extends from `Player`?

**Review by marchermans — CHANGES_REQUESTED** [review]

File: src/main/java/net/neoforged/neoforge/event/entity/player/ServerPlayerEvent.java (hunk 2 above)
Comment (Spinoscythe) [reply src/main/java/net/neoforged/neoforge/event/entity/player/ServerPlayerEvent.java]: I was thinking the same thing but I un-did it for some reason I can't remember

File: src/main/java/net/neoforged/neoforge/event/entity/item/ItemTossEvent.java [hunk 3]
```diff
@@ -21,11 +21,11 @@ public class ItemTossEvent extends ItemEvent implements ICancellableEvent {
     /**
      * Creates a new event for EntityItems tossed by a player.
      * 
-     * @param entityItem The EntityItem being tossed.
+     * @param itemEntity The EntityItem being tossed.
```
Comment (TelepathicGrunt) [inline src/main/java/net/neoforged/neoforge/event/entity/item/ItemTossEvent.java]: Don't forget this EntityItem here too and above. We call them ItemEntity as that is what the actual class is. Someone must had typo'ed EntityItem before

**Comment by Shadows-of-Fire** [comment]
This change doesn't really make sense.  If we're going to make breaking changes to these events, they should be moved to their own individual files. There isn't any reason for `ServerPlayerEvent` to exist.

**Comment by Spinoscythe** [comment]
> This change doesn't really make sense. If we're going to make breaking changes to these events, they should be moved to their own individual files. There isn't any reason for `ServerPlayerEvent` to exist.

I think having `ServerPlayerEvent` saves time from casting `getEntity` in every event to `ServerPlayer`

**Comment by Caltinor** [comment]
> > This change doesn't really make sense. If we're going to make breaking changes to these events, they should be moved to their own individual files. There isn't any reason for `ServerPlayerEvent` to exist.
> 
> I think having `ServerPlayerEvent` saves time from casting `getEntity` in every event to `ServerPlayer`

I noticed a code format difference.  In subclasses of `LivingEvent` which return a specific entity type (namely `PlayerEvent` and `AnimalTameEvent`), they each take the specific entity class for their event class name (`Player` and `Animal` respectively), pass that into super but also store it locally and pass it back via the `getEntity` override.  
```java
public class ExampleEntityEvent extends LivingEvent {
  private final ExampleEntity example;
  
  public ExampleEntityEvent(ExampleEntity example) {
    super(example);
    this.example = example;
  }

  @Override
  public ExampleEntity getEntity() {return example;}
}
```
This implementation in contrast just casts the super invocation.  Whether this or the other classes change, I think we should be consistent.

**Comment by TelepathicGrunt** [comment]
When you can, please implement [Caltinor](https://github.com/Caltinor)'s suggestion. It would remove the need for the casting in the server player events

**Comment by Caltinor** [comment]
I think this https://github.com/neoforged/NeoForge/pull/1153/commits/c7cf77793830c8527d2afba6bf51888f8a88bddd is the opposite of what @TelepathicGrunt wanted. They wanted to remove the casts, from how I read it.

**Comment by TelepathicGrunt** [comment]
Yeah. Less casting. Not more

**Comment by Shadows-of-Fire** [comment]
Per [a conversation in the NF discord](https://discord.com/channels/313125603924639766/1154167065519861831/1259399586473054271):

Introducing a generic to EntityEvent allows us to streamline more than just the server player events:  
We'd be able to drop `LivingEvent`, `ItemEvent`, and `PlayerEvent` just by having that generic there.  We introduce some redundancy where we may end up having `EntityEvent<Entity>` (which would be a nontrivial number of cases), but there's also a benefit: Authors of events that are children of `EntityEvent<T>` should now be thinking about what the correct `T` is for the situation (so we don't end up with events that are too lenient as we have now).

Based on that, this PR is not something we want to proceed with at this time, and the refactor triggered by switching to `EntityEvent<T>` is large enough it warrants its own new PR.

**Pull request closed without merging**